chalice/utils.py

```diff
@@ -29,10 +28,10 @@ def to_cfn_resource_name(name):
     This transform ensures that only alphanumeric characters are used
     and prevent collisions by appending the hash of the original name.
     """
-    alphanumeric_only_name = re.sub(r'[^A-Za-z0-9]+', '', name)
-    return ''.join([
-        alphanumeric_only_name, hashlib.md5(
-            name.encode('utf-8')).hexdigest()[:4]])
+    for word_separator in ['-', '_']:
+        word_parts = name.split(word_separator)
+        name = ''.join([w[0].upper() + w[1:] for w in word_parts])
```

Comment:
This breaks if you have a resource that starts with underscore (i.e. `_index()`). Do you think we need to account for this? I can see this being helpful.

Reply:
Yeah, I'll also add property tests for this as well.